backend/attachRendererFiber.js

```diff
@@ -0,0 +1,244 @@
+        continue;
+      }
+      while (node.return) {
+        node = node.return;
+        enqueueUnmount(node, events);
+        if (node == fiber) {
+          return;
+        }
+        if (node.sibling) {
+          node.sibling.return = node.return;
+          node = node.sibling;
+          continue outer;
+        }
+      }
+      return;
+    }
+  }
+
+  function mountFiber(fiber, events) {
+    // Depth-first.
+    // Logs mounting of children first, parents later.
+    let node = fiber;
+    outer: while (true) {
+      if (node.child) {
+        node.child.return = node;
+        node = node.child;
+        continue;
+      }
+      enqueueMount(node, events);
+      if (node == fiber) {
+        return;
+      }
+      if (node.sibling) {
+        node.sibling.return = node.return;
+        node = node.sibling;
+        continue;
+      }
+      while (node.return) {
+        node = node.return;
+        enqueueMount(node, events);
+        if (node == fiber) {
+          return;
+        }
+        if (node.sibling) {
+          node.sibling.return = node.return;
+          node = node.sibling;
+          continue outer;
+        }
+      }
+      return;
+    }
+  }
+
+  function updateFiber(nextFiber, prevFiber, events) {
+    // TODO: optimize for the common case of children with implcit keys.
+    // Just like we do in the Fiber child reconciler.
+    // We shouldn't be allocating Maps all the time.
+    const allKeys = new Set();
+    const prevChildren = mapChildren(prevFiber, allKeys);
+    const nextChildren = mapChildren(nextFiber, allKeys);
+    allKeys.forEach(key => {
+      const prevChild = prevChildren.get(key);
+      const nextChild = nextChildren.get(key);
+      if (prevChild != null && nextChild != null) {
+        // TODO: are there more cases when we can bail out?
+        if (prevChild !== nextChild) {
+          if (prevChild.type === nextChild.type) {
+            // We already know their keys match.
+            updateFiber(nextChild, prevChild, events);
+          } else {
+            // These are different fibers.
+            unmountFiber(prevChild, events);
+            mountFiber(nextChild, events);
+          }
+        }
+      } else if (nextChild != null) {
+        mountFiber(nextChild, events);
+      } else if (prevChild != null) {
+        unmountFiber(prevChild, events);
+      }
+    });
+    enqueueUpdate(nextFiber, events);
+  }
+
+  let subscription = null;
+
+  function cleanup() {
+    if (subscription) {
+      const {unsubscribe} = subscription;
+      subscription = null;
+      unsubscribe();
+    }
+  }
+
+  function emitEvents(events) {
+    for (let i = 0; i < events.length; i++) {
+      const event = events[i];
+      hook.emit(event._event, event);
+    }
+  }
+
+  function handleFiberCommit(root) {
+    const current = root.current;
+    const alternate = current.alternate;
+    const events = [];
+    if (alternate && subscription != null) {
```

Comment:
Then `subscription` would be null (since we haven't finished subscribing, and the React side calls the commit hook synchronously during subscription the first time).